OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs

```diff
@@ -0,0 +1,341 @@
+		{
+			this.self = self;
+		}
+
+		#region IDockHost
+
+		public BitSet<DockType> GetDockType => Info.Type;
+
+		public bool IsEnabledAndInWorld => !preventDock && !IsTraitDisabled && !self.IsDead && self.IsInWorld;
+		public int ReservationCount => ReservedDockClients.Count;
+		public bool CanBeReserved => ReservationCount < Info.MaxQueueLength;
+
+		public WPos DockPosition => self.CenterPosition + Info.DockOffset;
+		public int DockWait => Info.DockWait;
+		public WAngle DockAngle => Info.DockAngle;
+
+		public virtual bool IsDockingPossible(Actor clientActor, IDockClient client, bool ignoreReservations = false)
+		{
+			return !IsTraitDisabled && (ignoreReservations || CanBeReserved || ReservedDockClients.Contains(client.DockClientManager));
+		}
+
+		public virtual bool Reserve(Actor self, DockClientManager client)
+		{
+			if (CanBeReserved && !ReservedDockClients.Contains(client))
+			{
+				ReservedDockClients.Add(client);
+				client.ReserveHost(self, this);
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void UnreserveAll()
+		{
+			while (ReservedDockClients.Count > 0)
+				Unreserve(ReservedDockClients[0]);
+		}
+
+		public virtual void Unreserve(DockClientManager client)
+		{
+			if (ReservedDockClients.Contains(client))
+			{
+				ReservedDockClients.Remove(client);
+				client.UnreserveHost();
+			}
+		}
+
+		public virtual void OnDockStarted(Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (Info.OccupyClient || DockWait <= 0)
+				DockStarted(self, new TraitPair<DockClientManager>(clientActor, client));
+			else
+				WaitingClients.Add((new TraitPair<DockClientManager>(clientActor, client), DockWait + self.World.WorldTick));
+		}
+
+		public virtual void OnDockCompleted(Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (clientActor != null && !clientActor.IsDead && clientActor.IsInWorld)
+				client.OnDockCompleted(clientActor, self, this);
+
+			DockedClients.Remove(DockedClients.First(c => c.Trait == client));
+		}
+
+		public virtual bool QueueMoveActivity(Activity moveToDockActivity, Actor self, Actor clientActor, DockClientManager client)
+		{
+			if ((clientActor.CenterPosition - DockPosition).HorizontalLengthSquared > Info.Range.LengthSquared)
+			{
+				// TODO: MoveWithinRange doesn't support offsets.
+				moveToDockActivity.QueueChild(clientActor.Trait<IMove>().MoveWithinRange(Target.FromActor(self), Info.Range));
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void QueueDockActivity(Activity moveToDockActivity, Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (Info.OccupyClient)
+			{
+				if (moveToDockActivity == null)
+					clientActor.QueueActivity(new RemoteDockSequence(client, self, this, Info.Range.Length));
+				else
+					moveToDockActivity.QueueChild(new RemoteDockSequence(client, self, this, Info.Range.Length));
+			}
+			else
+			{
+				// When auto-docking, this function be repeadedly called.
```

Comment:
```suggestion
				// When auto-docking, this function will be called repeatedly.
```
?

Also this really shouldn't be handled by using `QueueDockActivity` but using another method, since we are not queuing a docking activity. (That or the method name needs to be adjusted.)

Reply:
This is an interface implementation, so I'm not sure wether this should be changed